solutions/haskell/07-expiry/diff/app/Main.hs.diff

```diff
@@ -0,0 +1,260 @@
+ import Data.ByteString (ByteString, pack)
+ import qualified Data.ByteString.Char8 as B
+ import Prelude hiding (concat)
+ import Text.Megaparsec
+     ( parse,
+       count,
+       (<|>),
+       Parsec,
+       MonadParsec(try),
+       Stream(Tokens) )
+ import Text.Megaparsec.Byte ( crlf, printChar, string )
+ import Text.Megaparsec.Byte.Lexer (decimal)
+ import Data.Void ( Void )
+ import Data.Text ( toLower, Text )
+ import Data.Text.Encoding (decodeUtf8)
+ import Data.Map (Map, insert, findWithDefault, empty)
+ import Data.Map.Internal.Debug (showTree)
+ import Control.Concurrent.STM (atomically, newTVarIO, readTVarIO, TVar)
+ import Control.Concurrent.STM.TVar (modifyTVar)
++import Data.Time
++    ( UTCTime,
++      defaultTimeLocale,
++      parseTimeOrError,
++      addUTCTime,
++      diffUTCTime,
++      getCurrentTime )
+
+ type Request = ByteString
+ type Response = ByteString
+ type Parser = Parsec Void Request
+ type Message = ByteString
+ type Key = ByteString
+ type Value = ByteString
+-type DB = Map Key Value
++type DB = Map Key (Value, Expiry)
++type Time = Maybe Integer
++type Expiry = UTCTime
+
+ data Command = Ping
+              | Echo Message
+-             | Set Key Value
++             | Set Key Value Time
+              | Get Key
+ data ApplicationError = UnknownCommand
+ data ProgConfigs = ProgConfigs {
+-    recvBytes :: Int }
++    recvBytes :: Int,
++    noExpiry :: UTCTime,
++    timeFormat :: String }
+ data CmdNames = CmdNames {
+     pingN :: Text,
+     echoN :: Text,
+     setN :: Text,
+     getN :: Text }
+ data RedisSpecs = RedisSpecs {
+     port :: String,
+     pingDefault :: Response,
+     unknownCmd :: Response,
+     setSuccess :: Response,
+     nilString :: Response,
+     bulkStringId :: ByteString,
+     arrayId :: ByteString,
+-    simpleStringId :: ByteString }
++    simpleStringId :: ByteString,
++    msOption :: Text,
++    nullStringId :: ByteString }
+
+ main :: IO ()
+ main = do
+     putStrLn $ "\r\n>>> Redis server listening on port " ++ port redisSpecs ++ " <<<"
+     redisDB <- setupDB
+     serve HostAny (port redisSpecs) $ \(socket, _address) -> do
+         putStrLn $ "successfully connected client: " ++ show _address
+         _ <- forever $ do
+             request <- recv socket $ recvBytes progConfigs
+             response <- do
+                 case parseRequest request of
+                     Left _ -> return $ unknownCmd redisSpecs
+                     Right cmd -> exec cmd redisDB
+             _ <- send socket (encodeRESP response)
+
+             -- debug database
+             out <- readTVarIO redisDB
+             putStrLn $ "\r\n***\r\nRedis DB content:\r\n"++ showTree out ++ "***\r\n"
+         putStrLn $ "disconnected client: " ++ show _address
+
+ progConfigs :: ProgConfigs
+ progConfigs = ProgConfigs {
+-                recvBytes  = 2048 }
++                recvBytes  = 2048,
++                noExpiry   = toUTCTime "2222-12-31T23:59:59",
```

Comment:
Is there a way to model this without using an artificial future timestamp? Something like `Expiry = Maybe UTCTime`?

Reply:
Good point! It took me a while to figure this one out, but I found a nice way to model this using the `Maybe` type. I am using the type `Maybe Expire` to indicate that the database stores something of type `Maybe` along with the time.